warehouse/warehouse.go

```diff
@@ -961,78 +1358,209 @@ func (wh *HandleT) setupTables() {
+		}
+		claim.ClaimResponseChan <- response
 	}
-	return err
+	slaveWorkerRoutineStatus[workerIdx-1] = false
+	logger.Infof("WH: Setting free slave worker %d: %v", workerIdx, slaveWorkerRoutineStatus)
 }
 
-func (wh *HandleT) Setup(whType string) {
-	logger.Infof("WH: Warehouse Router started: %s", whType)
+func setupSlave() {
+	slaveWorkerRoutineStatus = make([]bool, noOfSlaveWorkerRoutines)
+	slaveID := uuid.NewV4().String()
+	rruntime.Go(func() {
+		jobNotificationChannel, err := notifier.Subscribe("process_staging_file")
+		if err != nil {
+			panic(err)
+		}
+		for {
+			ev := <-jobNotificationChannel
+			logger.Infof("WH: Notification recieved, event: %v, workers: %v", ev, slaveWorkerRoutineStatus)
+			for workerIdx := 1; workerIdx <= noOfSlaveWorkerRoutines; workerIdx++ {
+				if !slaveWorkerRoutineStatus[workerIdx-1] {
+					slaveWorkerRoutineStatus[workerIdx-1] = true
+					idx := workerIdx
+					rruntime.Go(func() {
+						func(index int) {
+							claimAndProcess(index, slaveID)
+						}(idx)
+					})
+					break
+				}
+			}
+		}
+	})
+}
+
+func isStandAlone() bool {
+	return warehouseMode != EmbeddedMode
+}
+
+func isMaster() bool {
+	return warehouseMode == config.MasteMode || warehouseMode == config.MasterSlaveMode || warehouseMode == config.EmbeddedMode
+}
+
+func isSlave() bool {
+	return warehouseMode == config.SlaveMode || warehouseMode == config.MasterSlaveMode || warehouseMode == config.EmbeddedMode
+}
+
+func Start() {
+	// do not start warehouse service if rudder core is not in normal mode and warehouse is running in same process as rudder core
+	if !isStandAlone() && !db.IsNormalMode() {
+		logger.Infof("Skipping start of warehouse service...")
+		return
+	}
+
+	logger.Infof("WH: Starting Warehouse service...")
 	var err error
-	psqlInfo := jobsdb.GetConnectionString()
-	wh.dbHandle, err = sql.Open("postgres", psqlInfo)
+	psqlInfo := getConnectionString()
+
+	if !misc.IsPostgresCompatible(psqlInfo) {
+		err := errors.New("Rudder Warehouse Service needs postgres version >= 10. Exiting")
+		logger.Error(err)
+		panic(err)
+	}
+
+	dbHandle, err = sql.Open("postgres", psqlInfo)
 	if err != nil {
 		panic(err)
 	}
-	wh.setupTables()
-	wh.destType = whType
-	wh.setInterruptedDestinations()
-	wh.Enable()
-	wh.uploadToWarehouseQ = make(chan []ProcessStagingFilesJobT)
-	wh.createLoadFilesQ = make(chan LoadFileJobT)
-	rruntime.Go(func() {
-		wh.backendConfigSubscriber()
-	})
-	rruntime.Go(func() {
-		wh.initUploaders()
-	})
-	rruntime.Go(func() {
-		wh.initWorkers()
-	})
+	setupTables(dbHandle)
+	notifier, err = pgnotifier.New(psqlInfo)
+	if err != nil {
+		panic(err)
+	}
+
+	if isSlave() {
+		logger.Infof("WH: Starting warehouse slave...")
+		setupSlave()
+	}
+
+	if isMaster() {
+		backendconfig.Setup()
+		logger.Infof("WH: Starting warehouse master...")
+		err = notifier.AddTopic("process_staging_file")
+		if err != nil {
+			panic(err)
+		}
+		rruntime.Go(func() {
+			monitorDestRouters()
+		})
+	}
+
 	rruntime.Go(func() {
```

Comment:
We can just call startWebHandler? No need of an extra goroutine and keepProcessAlive?

Reply:
yes. written before slave had webhandler. removed it.